src/dict_list.jl

```diff
@@ -0,0 +1,210 @@
+                        end
+                        Dict([k=>lookup_candidate(trial,k) for k in keys(trial)])
+                    end)
+        partially_restricted_parameters = Set([k for k in keys(c) if 
+                                               contains_partially_restricted(c[k]) && !is_fully_restricted(c[k])])
+        return collect(filter(parameter_sets) do ps
+            partially_restricted_parameters ⊆ Set(keys(ps))
+        end)
+    end
+    return _dict_list(c)
+end
+
+function _dict_list(c::Dict)
+    iterable_fields = filter(k -> typeof(c[k]) <: Vector, keys(c))
+    non_iterables = setdiff(keys(c), iterable_fields)
+
+    iterable_dict = Dict(iterable_fields .=> getindex.(Ref(c), iterable_fields))
+    non_iterable_dict = Dict(non_iterables .=> getindex.(Ref(c), non_iterables))
+
+    vec(
+        map(Iterators.product(values(iterable_dict)...)) do vals
+            dd = Dict(keys(iterable_dict) .=> vals)
+            if isempty(non_iterable_dict)
+                dd
+            elseif isempty(iterable_dict)
+                non_iterable_dict
+            else
+                merge(non_iterable_dict, dd)
+            end
+        end
+    )
+end
+
+"""
+    dict_list_count(c) -> N
+Return the number of dictionaries that will be created by
+calling `dict_list(c)`.
+"""
+function dict_list_count(c)
+    if DependentParameter in eltype.(values(c))
+        return length(dict_list(c))
+    end
+    iterable_fields = filter(k -> typeof(c[k]) <: Vector, keys(c))
+    prod(length(c[i]) for i in iterable_fields)
+end
+
+# Basis taken from https://github.com/FluxML/MacroTools.jl
+walk(x, inner, outer, ex) = outer(x, ex)
+walk(x::Expr, inner, outer, ex) = outer(Expr(x.head, map(y->inner(y,x), x.args)...), ex)
+postwalk(f, x, ex=:()) = walk(x, (x,y) -> postwalk(f, x, y), f, ex)
+
+export @onlyif
+
+struct DependentParameter{T}
+    value::T
+    condition::Function
+end
+
+contains_partially_restricted(d::Dict) = any(contains_partially_restricted,values(d))
+contains_partially_restricted(d::Vector) = any(contains_partially_restricted,d)
+contains_partially_restricted(::DependentParameter) = true 
+contains_partially_restricted(::Any) = false 
+contains_fully_restricted(d::Dict) = any(is_fully_restricted,values(d))
+is_fully_restricted(d::Vector) = eltype(d) <: DependentParameter
+is_fully_restricted(d) = typeof(d) <: DependentParameter
+
+function toDependentParameter(value::T,condition) where T
+    if T <: Vector
+        return DependentParameter.(value,Ref(condition))
+    end
+    return DependentParameter(value,condition)
+end
+
+function lookup_candidate(d, name)
+    if name in keys(d)
+        if d[name] isa DependentParameter
+            return d[name].value
+        end
+        return d[name]
+    end
+    return name
+end
+
+"""
+    @onlyif(ex, value)
+Tag `value` to only appear in a dictionary created with [`dict_list`](@ref) if
+the Julia expression `ex` is evaluated as true.  If `value` is a subtype of
```

Comment:
Hm, here I was confused about what `ex` is.. I thought it would be a boolean value, but it is actually a Julia expresision containing previous keys of the `dict_list` argument, represented as symbols, and operations involving them. Perhaps we should say this.

Reply:
The examples of course clarify this as well.

Reply:
Ah, forget this whole coversation, I see you explain it perfectly below.